src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs

```diff
@@ -874,6 +874,403 @@ string M(object o, bool b)
+{
+    void M(object e)
+    {
+        for (; !(!(e is C c));) { }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestForStatement_InScope()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    void M(object e)
+    {
+        [|C|] c = null;
+        for (; !((c = e as C)==null);)
+        {
+            M(c);
+        }
+    }
+}",
+@"class C
+{
+    void M(object e)
+    {
+        for (; !(!(e is C c));)
+        {
+            M(c);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestForStatement_NotAssignedBeforeAccess()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class C
+{
+    void M(object e)
+    {
+        [|C|] c = null;
+        for (; ((c = e as C)==null);)
+        {
+            if (b) c = null;
+            M(c);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestForStatement_AssignedBeforeAccess()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    void M(object e, bool b)
+    {
+        [|C|] c = null;
+        for (; (c = e as C)==null;)
+        {
+            if (b) c = null;
+            else c = null;
+            M(c);
+        }
+    }
+}",
+@"class C
+{
+    void M(object e, bool b)
+    {
+        for (; !(e is C c);)
+        {
+            if (b) c = null;
+            else c = null;
+            M(c);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestForStatement_MultipleDeclarators()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    void M(object e)
+    {
+        [|C|] c = null, x = null;
+        for (; !((c = e as C)==null);)
+        {
+            M(c);
+        }
+    }
+}",
+@"class C
+{
+    void M(object e)
+    {
+        C x = null;
+        for (; !(!(e is C c));)
+        {
+            M(c);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestForStatemnet_UseBeforeDeclaration()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class C
+{
+    void M(object e)
+    {
+        [|C|] c = null, x = c;
+        for (; !((c = e as C)==null);)
+        {
+            M(c);
+        }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task TestLocalFunction()
+        {
+            await TestInRegularAndScriptAsync(
+@"class C
+{
+    void M(object e)
+    {
+        [||]var c = e as C;
+        C F() => c == null ? null : c;
+    }
+}",
+@"class C
+{
+    void M(object e)
+    {
+        C F() => !(e is C c) ? null : c;
```

Comment:
hrmm... should we warn if you cross a lambda/function scope?

Reply:
we will check if the variable is used out of new scope, so we don't have a special case